### DISCUSSION TITLE
maybe rect  strokeUniform not work will

### DISCUSSION THREAD

**Discussion by vaynevayne** [fabricjs/fabric.js#10307] [General]
`STROKE_WIDTH` Both are 1. In 4.5, use zoom to zoom out the picture, and the rectangular border will be clear.
In v6, the rectangular border is not clear,
setViewport is used in v6
In v4, only setZoom is used
It feels like `strokeUniform` is ineffective

fabric 4.5.0 
![image](https://github.com/user-attachments/assets/a880389b-dd83-467d-b391-add136c25fc6)
```jsx
const STROKE_WIDTH =1
const base_config = {
    width: isOuter ? w - STROKE_WIDTH : w,
    height: isOuter ? h - STROKE_WIDTH : h,
    left: x + (scaleObj.left || 0),
    top: y + (scaleObj.top || 0),
    stroke: '#04aa65',
    strokeWidth: STROKE_WIDTH,
    fill: 'transparent',
    class: CLASS_RECT,
    // 将元素的 strokeUniform 属性设置为 true 后，元素在缩放后，边框会恢复成最初设置的大小。
    strokeUniform: true,
    data,
    ...(data.config || {}),
    // strokeUniform: true,
  }
```

fabric 6.4.3
![image](https://github.com/user-attachments/assets/0c616247-1399-4441-844e-133c45900aa8)

```jsx
{
    "rx": 0,
    "ry": 0,
    "type": "Rect",
    "version": "6.5.0",
    "originX": "left",
    "originY": "top",
    "left": -1430.3856,
    "top": -305.0278,
    "width": 1429.3856,
    "height": 790.7743,
    "fill": "transparent",
    "stroke": "#04aa65",
    "strokeWidth": 1,
    "strokeDashArray": null,
    "strokeLineCap": "butt",
    "strokeDashOffset": 0,
    "strokeLineJoin": "miter",
    "strokeUniform": true,
    "strokeMiterLimit": 4,
    "scaleX": 1,
    "scaleY": 1,
    "angle": 0,
    "flipX": false,
    "flipY": false,
    "opacity": 1,
    "shadow": null,
    "visible": true,
    "backgroundColor": "",
    "fillRule": "nonzero",
    "paintFirst": "fill",
    "globalCompositeOperation": "source-over",
    "skewX": 0,
    "skewY": 0
}
```

**Comment by vaynevayne** [comment]
@zhe-he Hello, do you have any ideas based on the known information?
I don't have time to provide a recurring demo yet, when I can provide more information

**Reply by zhe-he** [reply]
Just turn off caching.
FabricObject.ownDefaults.objectCaching = false;
or
obj.objectCaching = false;

[demo](https://codepen.io/zhe-he-the-vuer/pen/NPKqwad)

**Reply by vaynevayne** [reply]
@zhe-he  not work 
```jsx
import { Rect as FabricRect } from 'fabric6'
FabricRect.ownDefaults.objectCaching = false

 <Rect {...rect} objectCaching={false} />
 
```
I looked at your demo and didn't see anything different.

**Reply by zhe-he** [reply]
Show your question with a minimal example, preferably with an online link.

**Reply by vaynevayne** [reply]
https://codesandbox.io/p/github/vaynevayne/react-fabric/main?import=true @zhe-he 

The point given is a little big, please scroll the scroll wheel, shrink the canvas, and you will see it.
<img width="1166" alt="image" src="https://github.com/user-attachments/assets/41fbf2e2-981b-474f-b430-9b7412e0fb50">

**Reply by zhe-he** [reply]
The following two are the differences between not adding objectCaching={false} and adding it. It does take effect, doesn't it?

```ts
<Rect key={index} {...rect}  left={rect.defaultLeft} strokeWidth={1} top={rect.defaultTop}  width={rect.defaultWidth} height={rect.defaultHeight}/>
```
```ts
<Rect key={index} {...rect}  objectCaching={false}  left={rect.defaultLeft} strokeWidth={1} top={rect.defaultTop}  width={rect.defaultWidth} height={rect.defaultHeight}/>
```


https://github.com/user-attachments/assets/fdcf3e01-8491-4121-8b92-3eb6aadcd618

https://github.com/user-attachments/assets/3a4a7f71-d69c-4f53-b42c-391599ba9f18

**Reply by zhe-he** [reply]
When scaling the shape, `strokeUniform` takes effect.  [demo](https://codepen.io/zhe-he-the-vuer/pen/NPKqwad)


https://github.com/user-attachments/assets/f9a2d95a-7067-44a5-a362-30c727aa8fd9

**Reply by vaynevayne** [reply]
`objectCaching={false} ` It takes effect, but the border still looks very thin. Is this expected behavior?
I want it to look the same as version 4.5, is there any solution?

4.5 version, i use api not react-fabric,  I zoom the canvas and background image first, and    Promise.then draw a rect,
In version 6.x, I drew rectangles and zoomed separately
Is that why? @zhe-he

**Reply by zhe-he** [reply]
> objectCaching={false}  It takes effect, but the border still looks very thin. Is this expected behavior?

Yes
No matter how you scale the canvas, the border size of the shape will not change. The value of strokeUniform is related to the scaling of the shape itself.

I implemented the same demo using version 4.5.0 and found that they exhibit the same behavior.
[4.5.0 demo](https://codepen.io/zhe-he-the-vuer/pen/jENqjBg?editors=0010)

**Reply by vaynevayne** [reply]
<img width="847" alt="image" src="https://github.com/user-attachments/assets/40de4d09-75a8-4d59-b753-b26148f0916e">
This is the 4.5 frame method, calculate the scale and then frame the frame.
You mean that it has nothing to do with the order, zoom will not cause the border to fade?
So what caused this problem? How should I solve it? What do you suggest? @zhe-he

**Reply by zhe-he** [reply]
The canvas scaling I’m discussing refers to ctx scaling, whereas the scaling you mentioned appears to be CSS scaling. If the border size change is caused by CSS scaling, you might need to proportionally increase the border size of the shape to compensate.
Maybe I misinterpreted it; after all, the code is just a snippet, so I can only make guesses.

**Comment by vaynevayne** [comment]
@asturur  Hello, do you have time to help take a look?

**Reply by asturur** [reply]
i didn't have yet, i have to many emails and open issues

**Comment by vaynevayne** [comment]
https://github.com/fabricjs/fabric.js/discussions/10323
I see an image that is blurry because the zoom factor is too small. Is this the reason why the border is blurry? @zhe-he
